### ISSUE TITLE
TextureMesh : Texture Re-generation

### ISSUE THREAD

**Issue by thunders82** [cdcseacave/openMVS#459]
Hi,

Is it possible to re-create the texture without re-creating new the UV's / atlas? I'd like to re-use the same UVs as the 1st texture generation done with TextureMesh.

Thank you,

Geoffrey

step 1 : use TextureMesh to generate textured mesh with UVs
step 2 : delete or rename texture
step 3 : use TextureMesh to re-generate the texture only (without modifying the mesh and UV's)

**Comment by cdcseacave** [comment]
anything is possible but as the code is not there is not an easy way
are you interested in using different images?

**Comment by thunders82** [comment]
Yes that would be really useful. I was thinking to save the 'atlas/mapping' into a text file and reuse it later. I'm not sure if it would work but you may have a better idea?
I'm not a heavy coder... More a script guy but I can give it a go if you guide me?

**Comment by cdcseacave** [comment]
what u need to save is the image id used for each triangle in the mesh
second time you run, do not recompute that, just load them and apply them on the already existent patches
the problem however is that in general the other images you plan to use are not taken with the exact same camera model and exact pose as the original images; that will result in ghosting, missalignment and many other artifact in the new texture; in order to support that, a new bundle adjustment needs to be run on the new images to correct all those above

**Comment by thunders82** [comment]
Hi,
I had a quick look at the code. Does it mean I have to save 'texturePatches' and 'rects' ?

Thanks

**Comment by thunders82** [comment]
Hi!

I had a go at it and managed to save the needed data (could you confirm ? ) : texturePaches.label , texturePatches.rect and rects.

When I re-apply those values I have got the exact same texture layout and similar color output. We can notice some small discrepancies in terms of colour (missing texture blending/colour grading?)

my modified code is only using : 

```
{
		TD_TIMER_STARTD();
		if (!texture.FaceViewSelection(fOutlierThreshold, fRatioDataSmoothness))
			return false;
		DEBUG_EXTRA("Assigning the best view to each face completed: %u faces (%s)", mesh.faces.GetSize(), TD_TIMER_GET_FMT().c_str());
	}
```

and my modified : 
```
{
		TD_TIMER_STARTD();
		texture.ReGenerateTexture(bGlobalSeamLeveling, bLocalSeamLeveling, nTextureSizeMultiple, nRectPackingHeuristic, colEmpty);
		DEBUG_EXTRA("Re-Generating texture atlas and image completed: %u image size (%s)", mesh.textureDiffuse.width(), TD_TIMER_GET_FMT().c_str());
	}
```


And here is the code of my ReGenerateTexture : 


```
void MeshTexture::ReGenerateTexture(bool bGlobalSeamLeveling, bool bLocalSeamLeveling, unsigned nTextureSizeMultiple, unsigned nRectPackingHeuristic, Pixel8U colEmpty)
{
  // create texture image
  ifstream file("patches.txt");
  if (file.is_open()) {

    int size,patches;
    uint32_t mylabel;
    string line;
    vector<string> list;
    file >> size >> patches;

    // cout << "TEXTURE SIZE : " << size << endl;
    // cout << "PATCH COUNT : " << patches << endl;

    textureDiffuse.create(size, size);
    textureDiffuse.setTo(cv::Scalar(colEmpty.b, colEmpty.g, colEmpty.r));

    for (int i=0; i<(patches-1); ++i) {

      // GET LABEL
      cout << i ;
      file>>mylabel;
      cout << " : " << mylabel;
      // GET TEXTURE PATCH RECT
      file>>line;
      boost::split(list, line, boost::is_any_of(","));
      const RectsBinPack::Rect texturePatchRect(stoi(list[0]),stoi(list[1]),stoi(list[2]),stoi(list[3]));
      cout << " " << texturePatchRect ;
      //GET RECT
      file>>line;
      boost::split(list, line, boost::is_any_of(","));
      const RectsBinPack::Rect rect(stoi(list[0]),stoi(list[1]),stoi(list[2]),stoi(list[3]));
      cout << " " << rect << endl;
      // copy patch image
      ASSERT((rect.width == texturePatchRect.width && rect.height == texturePatchRect.height) ||
      (rect.height == texturePatchRect.width && rect.width == texturePatchRect.height));
      int x(0), y(1);
      if (mylabel != NO_ID) {
        cv::Mat patch(images[mylabel].image(texturePatchRect));
        if (rect.width != texturePatchRect.width) {
          // flip patch and texture-coordinates
          patch = patch.t();
          x = 1; y = 0;
        }
        patch.copyTo(textureDiffuse(rect));
      }
   }
  }else {
    cout << "Couldn't open patches.txt" << endl;
  }
  file.close();
}
```



Any idea what could cause the colour discrepancies ?

Thank you,

Geoffrey

**Comment by cdcseacave** [comment]
the approach looks right, so it is possible that the new images do not perfectly align with the original images

**Comment by thunders82** [comment]
In this test I used exactly the same images/data. That's why it's a bit weird.

// create texture,UVs and save data in text file
$OPENMVS_PATH/TextureMesh -i ./scene_mesh_decim.mvs -o ./scene_mesh_decim_tex.mvs --cost-smoothness-ratio 1 --outlier-threshold 0.06
// reuse the same data and use text file to regenerate texture
$OPENMVS_PATH/ReTextureMesh -i ./scene_mesh_decim.mvs -o ./scene_mesh_decim_retex.mvs --cost-smoothness-ratio 1 --outlier-threshold 0.06

is it possible the colour grading/blending is re-evaluated and therefore could cause the discrepancies ?

**Comment by cdcseacave** [comment]
the blending is recomputed, but if the input to that function is exactly the same it should produce the exact same results

**Comment by thunders82** [comment]
I think I understand the problem... In my current code I'm not baking texturePatches.faces but it's used in the blending/color grading process. So I guess I need to bake it and reuse it too...
Now I've got some difficulties to understand how texturePatches.faces is structured. What would be the bast way to save it to file and reuse it ? Any hints ?

Thanks

**Comment by cdcseacave** [comment]
it is simply a vector of values
you can save it in many ways, just binary save to file, or boost::serialize (see example across the code)

**Comment by thunders82** [comment]
I'm currently doing it like this but wasn't sure it was correct as the output is HEX :

```
FOREACHPTR(pIdxFace, texturePatch.faces) {
    			myfile << pIdxFace << ",";
    		}
```
and from there I wasn't able to convert back the string from text file to uint32_t hex...

I'll check boost::serialize. Thank you

**Comment by thunders82** [comment]
While serialize works like a charm I realized I was stuck with the computation of the blending/color grading. I checked the 1st message you sent me :

> what u need to save is the image id used for each triangle in the mesh
> second time you run, do not recompute that, just load them and apply them on the already existent patches

I believe the image id for each triangle is computed  by FaceViewSelection() ?
and the data lives in : 
Mesh::FaceIdxArr components; // for each face, stores the texture patch index to which belongs
??

I tried to bake it just after FaceViewSelection() and re-used it in the second run without computing FaceViewSelection() but it's crashing.

As you can see I'm a bit lost... 
What are your thoughts ?

**Comment by thunders82** [comment]
Hi @cdcseacave 

I'm back from holidays and managed to make it work. I had to bake/save/load multiple variables at 2 different places in SceneTexture.cpp : 

1) Before "if (bGlobalSeamLeveling) " : 
-components
-mapIdxPatch
-seamVertices
-texturePatches
-scene

2) After "textureDiffuse.setTo(cv::Scalar(colEmpty.b, colEmpty.g, colEmpty.r));" : 
-rects
-texturePatches

It's quite hacky but it works. If there is any simpler way to do it please let me know. 

Here is a patch file if anyone is interested :  [file0001.patch.txt](https://github.com/cdcseacave/openMVS/files/3431341/file0001.patch.txt)

Now I'd like to use 32bit images for texturing (only). Anything I should know before starting or it should be straight forward? (let me know if I should open a new ticket)

Thank you,

Geoffrey

**Comment by cdcseacave** [comment]
Thx Geoffrey, great work.

There is currently no support for 32bit images, but if you patch OpenMVS to load them it should be relatively easy to make it work farther.

**Comment by thunders82** [comment]
Perfect. I ll have a look at the code and create a new ticket with any questions related to this.
I'm now closing this ticket.

Cheers,

Geoffrey.

**Comment by thunders82** [comment]
Hi @cdcseacave ,

Didn't noticed your last sentence : 'Pls make one Pull Request for each patches.'
I'll check how it works as I never did it. I have to admit I never used git as it should be either... 

I'll give it a go with this current patch and see how it goes.

Cheers,

Geoffrey

**Comment by thunders82** [comment]
HI @cdcseacave ,
I sent the pull request. Please let me know if it's done as expected.

Thanks
